src/storage/src/source/kafka.rs

```diff
@@ -185,6 +189,60 @@ impl SourceRender for KafkaSourceConnection {
         let (mut health_output, health_stream) = builder.new_output();
         let (mut stats_output, stats_stream) = builder.new_output();
 
+        let mut outputs = vec![];
+        for (
+            id,
+            SourceExport {
+                ingestion_output,
+                details,
+                storage_metadata: _,
+            },
+        ) in &config.source_exports
+        {
+            let resume_upper = Antichain::from_iter(
+                config
+                    .source_resume_uppers
+                    .get(id)
+                    .expect("all source exports must be present in source resume uppers")
+                    .iter()
+                    .map(Partitioned::<RangeBound<PartitionId>, MzOffset>::decode_row),
+            );
+
+            // Output index 0 is the primary source which is a valid output
+            // for legacy Kafka sources.
+            // TODO: When we remove support for outputting to the primary collection of
```

Comment:
Is there a world where this distinction only exists in the SQL layer? What prevents us from not having any of these special cases in the cluster code? As far as it is concerned it could be outputting multiple outputs and we just wire up the primary source as one of the exports here.

Reply:
Yep that's exactly where I'm planning to go with this -- one of the next few PRs is going to move the `encoding` and `envelope` specifiers from the source-connection to each `SourceExport`, and when I do that refactoring I was planning to fully populate the `SourceExport` for the primary-output such that it is indistinguishable from all other exports.

The main reason we can't do so immediately is that the sql layer currently populates the `details` as `SourceExportDetails::None` for the primary output since all of its 'details' are stored on the source connection itself. With the future PR the source-exports would all look the same and would likely be something like:
```
SourceExport {
    ingestion_output: int,
    details: SourceExportDetails::SourceType { source-export-specific-details },
    storage_metadata: _,
    envelope: SourceEnvelope,
    encoding: Option<SourceDataEncoding>,
},
```

Reply:
exciting! That sounds so good!